quickwit/quickwit-ingest/src/ingest_v2/rate_limiter.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::{Duration, Instant};
+
+use bytesize::ByteSize;
+use quickwit_common::tower::{ConstantRate, Rate};
+
+#[derive(Debug, Clone, Copy)]
+pub struct RateLimiterSettings {
+    pub burst_limit: ByteSize,
+    pub rate_limit: ConstantRate,
+    pub refill_period: Duration,
+}
+
+impl Default for RateLimiterSettings {
+    fn default() -> Self {
+        // 10 MB burst limit.
+        let burst_limit = ByteSize::mb(10);
+        // 5 MB/s rate limit.
+        let rate_limit = ConstantRate::bytes_per_sec(ByteSize::mb(5));
+        // Refill every 100ms.
+        let refill_period = Duration::from_millis(100);
+
+        Self {
+            burst_limit,
+            rate_limit,
+            refill_period,
+        }
+    }
+}
+
+/// A bursty token-based rate limiter.
+#[derive(Debug, Clone)]
+pub(super) struct RateLimiter {
+    capacity: u64,
+    available: u64,
+    refill_amount: u64,
+    refill_period: Duration,
+    refill_period_micros: u64,
+    refill_at: Instant,
+}
+
+impl RateLimiter {
+    pub fn from_settings(settings: RateLimiterSettings) -> Self {
+        let capacity = settings.burst_limit.as_u64();
+
+        let work = settings.rate_limit.work() as u128;
+        let refill_period = settings.refill_period;
+        let rate_limit_period = settings.rate_limit.period();
+        let refill_amount = work * refill_period.as_nanos() / rate_limit_period.as_nanos();
+
+        Self {
+            capacity,
+            available: capacity,
+            refill_amount: refill_amount as u64,
+            refill_period,
+            refill_period_micros: refill_period.as_micros() as u64,
+            refill_at: Instant::now() + refill_period,
+        }
+    }
+
+    pub fn acquire(&mut self, capacity: ByteSize) -> bool {
+        if self.acquire_inner(capacity.as_u64()) {
+            true
+        } else {
+            self.refill(Instant::now());
+            self.acquire_inner(capacity.as_u64())
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.available = self.capacity;
+        self.refill_at = Instant::now() + self.refill_period;
+    }
```

Comment:
is this used anywhere?

Reply:
Yes. It was useful before I introduced `RateLimiterSettings`, but I probably can remove it now.